lib/rbac/filterer.rb

```diff
@@ -317,23 +320,32 @@ def search(options = {})
       end
 
       if search_filter && targets && (!exp_attrs || !exp_attrs[:supported_by_sql])
-        rejects     = targets.reject { |obj| matches_search_filters?(obj, search_filter, tz) }
-        auth_count -= rejects.length unless options[:skip_counts]
-        targets -= rejects
+        targets = targets.lazy.select { |obj| matches_search_filters?(obj, search_filter, tz) }
+        unless options[:skip_counts]
+          # NOTE: the filter is not supported by sql (also means limit will be applied in ruby, too)
+          #       so: our count is calculated locally in ruby
+          #       also: use to_a to undo the lazy and filter one time before the count and applying the limit
+          targets    = targets.to_a
+          auth_count = targets.length
+        end
       end
+      attrs[:auth_count] = auth_count unless options[:skip_counts]
 
       if limit && !attrs[:apply_limit_in_sql]
-        attrs[:target_ids_for_paging] = targets.collect(&:id) # Save ids of targets, since we have then all, to avoid going back to SQL for the next page
-        offset = offset.to_i
-        targets = targets[offset...(offset + limit.to_i)] || []
+        unless options[:skip_counts]
+          # developer note: when changing code, ensure targets is not a lazy enumeration
+          # When we are applying limits in sql, that means we brought back every row.
+          #      we save the ids of the targets to avoid downloading them all again for the next page.
+          attrs[:target_ids_for_paging] = targets.collect(&:id)
+        end
+        targets = targets.drop(offset.to_i).take(limit.to_i)
       end
 
       # Preserve sort order of incoming target_ids
       if !target_ids.nil? && !order
         targets = targets.sort_by { |a| target_ids.index(a.id) }
       end
-
-      attrs[:auth_count] = auth_count unless options[:skip_counts]
+      targets = targets.to_a if targets.kind_of?(Enumerator::Lazy)
```

Comment:
it can be a scope, and in that case, we don't want to do a `to_a`

Reply:
@kbrock is this resolved?

Reply:
sorry, just saw. lazy will keep on evaluating every time we access the data. so if we do a size twice, it will run the miq expressions every time (2 times). Also enumeration is not quite an array, and we will need to add to_a in a few more places

Reply:
I documented why we need to drop the lazy down below. 📖